lib/persistent_enum.rb

```diff
@@ -92,20 +92,33 @@ class << self
     # by name using `ALTER TYPE` before insertion. This ensures that enum table
     # ids will have predictable values and can therefore be used in database
     # level constraints.
-    def cache_constants(model, required_members, name_attr: :name, required_attributes: nil, sql_enum_type: nil)
+    def cache_constants(model, required_members, name_attr: "name", required_attributes: nil, sql_enum_type: nil)
       # normalize member specification
       unless required_members.is_a?(Hash)
         required_members = required_members.each_with_object({}) { |c, h| h[c] = {} }
       end
 
+      # Normalize symbols
+      name_attr = name_attr.to_s
+      required_members = required_members.each_with_object({}) do |(name, attrs), h|
+        h[name.to_s] = attrs.transform_keys(&:to_s)
+      end
+      required_attributes = required_attributes.map(&:to_s) if required_attributes
+
       # We need to cope with (a) loading this class and (b) ensuring that all the
       # constants are defined (if not functional) in the case that the database
       # isn't present yet. If no database is present, create dummy values to
       # populate the constants.
       if model.table_exists?
-        table_attributes = (model.attribute_names - ["id", name_attr.to_s])
-        if required_attributes
-          table_attributes &= required_attributes.map(&:to_s)
+        internal_attributes = ["id", name_attr]
+        table_attributes = (model.attribute_names - internal_attributes)
+
+        # If not otherwise specified, only attributes without defaults are required
+        required_attributes ||= model.column_defaults.map { |attr, default| attr if default.nil? }.compact - internal_attributes
```

Comment:
What does this do in the presence of `default NULL` columns? It reads like it should consider them required.

Reply:
It does, which is not desirable. But Rails doesn't seem to collect that information, so it's not easy to detect only the no-default `NOT NULL` case.